### PR TITLE
treat invalidation as navigation

### LINKED ISSUE #4383: `beforeNavigate` fires when the page is reloaded
### Describe the bug

Previously, I was using the `sveltekit:navigation-start` and `sveltekit:navigation-end` events that were fired on `window` to show/hide a navigation loader on the page.

Now, it seems like the aforementioned events [were removed](https://github.com/sveltejs/kit/blob/master/packages/kit/CHANGELOG.md#100-next285) in favor of the `beforeNavigate`/`afterNavigate` lifecycle functions.

However, I expected the `beforeNavigate` lifecycle event to get fired only when client-side routing is triggered, just like how the good old `sveltekit:navigation-start` used to work, but I noticed that it gets fired even when you refresh the page, or manually modify the URL in the browser address bar and hit enter to leave the page.
https://github.com/sveltejs/kit/blob/e28bc390e033e38585fd1640da65c22f973afd2c/packages/kit/src/runtime/client/router.js#L110-L119
I don't understand the reason behind this, if someone wants to listen for `beforeunload` they could do so by simply registering a listener for that event; correct me if I'm wrong but there's no need for SvelteKit to additionally fire `beforeNavigate`, as that can be completely undesirable in certain cases, such as in the case of a client-side navigation loader, and it's also not consistent with the behavior of the old `sveltekit:navigation-start` event.
My primary suggestion is that the behavior should change to match that of the `sveltekit:navigation-start` event, namely that it should only be fired when client-side navigation occurs. This is the most expected behavior.
But at the very least, I think the callback function should be passed a boolean argument that indicates whether the navigation was triggered by the client-side router or not:
```js
beforeNavigate(({ client }) => {
    if (client) {
      // Blah blah blah...
    }
});
```

It's also worth noting that the `$navigating` store works as expected in this regard, namely that it gets populated only in the case of client-side routing. So, the current behavior of `beforeNavigate` is also inconsistent with the `$navigating` store.

### Logs

_No response_

### System Info

```shell
System:
    OS: Windows 10 10.0.19044
    CPU: (4) x64 Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz
    Memory: 885.61 MB / 7.89 GB
  Binaries:
    Node: 14.15.0 - C:\Program Files\nodejs\node.EXE     
    Yarn: 1.22.10 - ~\AppData\Roaming\npm\yarn.CMD       
    npm: 6.14.8 - C:\Program Files\nodejs\npm.CMD        
  Browsers:
    Edge: Spartan (44.19041.1266.0), Chromium (99.0.1150.39)
    Internet Explorer: 11.0.19041.1566
  npmPackages:
    @sveltejs/kit: ^1.0.0-next.301 => 1.0.0-next.301
    svelte: ^3.46.4 => 3.46.4
```

### Severity

serious, but I can work around it

### Additional Information

_No response_

### LINKED ISSUE #6375: `beforeInvalidate` and `afterInvalidate` events
### Describe the problem

I have a horizontal loader that's shown on every navigation that happens client-side (much like the one in kit.svelte.dev). It's using the `beforeNavigate` and `afterNavigate` events to appear and then disappear.

Occasionally I need to call `invalidate()` to re-run all the load functions (because the authentication status has changed, for example). I would like to show the same loader for when the load functions are re-running as well — because when you think about it, running `invalidate()` is kind of life refreshing the page.

### Describe the proposed solution

It would be nice if SvelteKIt provided two other lifecycle functions called `beforeInvalidate` and `afterInvalidate` as well, or perhaps integrated the invalidation process into the same `beforeNavigate` and `afterNavigate` functions and then adjust the naming.

### Alternatives considered

_No response_

### Importance

nice to have

### Additional Information

_No response_

### LINKED ISSUE #6498: has a popState
### Describe the problem

I'm porting a react webapp to svelte, and I have a need to know when a page was loaded by push/replace or because the user pressed the browser's back button or if window.history.go(-1) method was called.

For this, I'm using my own code (below), but I would like to know if sveltekit could officially inform that.

### Describe the proposed solution

This is working for my use case, 
but I'm sure it doesn't work for all cases

```js
import { browser } from '$app/environment'
import { afterNavigate } from '$app/navigation'

let idx = 0

export function listenPopState() {
	afterNavigate(() => {
		idx = window.history.state['sveltekit:index']
	})
}

export function isPopState() {
	return browser && idx > window.history.state['sveltekit:index']
}
```

### Alternatives considered

_No response_

### Importance

nice to have

### Additional Information

_No response_

### LINKED ISSUE #6517: Navigation types
### Describe the problem

Relevant: #4447, #4383, #6498.

The `navigating` store value is a `{ from: URL, to: URL }` object, which is also passed to `beforeNavigate` (along with a `cancel` method) and `afterNavigate` callbacks. 

Often, it's useful to know what _type_ of navigation is happening — whether we're coming from or going to a different document, or we navigated within the site to a different route by clicking on a link, or via `goto`, or by a `popstate` event (i.e. traversing the `history`). If it's a history traversal, it's also useful to know the delta (or at least whether we're going forward or back).

### Describe the proposed solution

Since the types are mutually exclusive, it probably makes sense to be an enum, rather than

```
{ load: true, goto: false, ... }
```

or something of that nature. Delta can be reported in a separate property that would only apply in the `popstate` case:

```js
beforeNavigate(({ from, to, type, delta }) => {
  if (delta) {
    console.log(`travelling ${delta < 0 ? 'backwards' : 'forwards'} through time`);
  } else switch (type) {
    case 'unload': console.log('bye for now'); break;
    case 'click': console.log('clickety clack'); break;
    case 'goto': console.log('as you wish'); break;
  }
});
```

(`load` isn't pictured because it would never be fired in the `beforeNavigate` case, just like `unload` would never appear in `afterNavigate`.)

### Alternatives considered

_No response_

### Importance

would make my life easier

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by Rich-Harris**
Update: since #6537 was merged, the only change this PR makes is to treat invalidation is equivalent to navigation for the purposes of `beforeNavigate`, `afterNavigate` and `navigating`. Leaving it here until we make a final decision, but there's some valid scepticism about whether this is the right move.

Original PR message follows.

---

closes #6517
closes #4383
closes #6498
closes #6375 
possibly #4447 too?

This adds a `type` property to the `Navigation` object found in various places:

```js
import { beforeNavigate, afterNavigate } from '$app/navigation';
import { navigating } from '$app/stores';

beforeNavigate(({ from, to, type }) => {
  console.log('beforeNavigate', type);
});

afterNavigate(({ from, to, type }) => {
  console.log('afterNavigate', type);
});

$: console.log($navigation?.type);
```

The type is one of the following:

* `load` — we just got here for the first time. `from` is `null` in this case
* `link` — the user clicked a link
* `goto` — the app called `goto` programmatically
* `popstate` — the user hit the back/forward button, or we're traversing the history some other way
* `invalidation` — the app called `invalidate(...)` or `invalidateAll()`. **This is a breaking change** — invalidations weren't previously treated as navigations
* `unload` — the page is being unloaded because the user clicked an external link, or submitted a form, or reloaded the page, etc. `to` will be `null` in this case

In the `popstate` case, there's an additional `delta` property, where -1 means navigating back and +1 means navigating forwards (but this can be any non-zero integer).

Todo

* [ ] docs

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. All changesets should be `patch` until SvelteKit 1.0

### CODE REVIEW

**Comment by aradalvand** [comment]
Correct me if I'm wrong but I believe this also closes #6375

**Comment by Rich-Harris** [comment]
Yep, you're right — nice

**Review by dummdidumm — APPROVED** [review]
All good implementation-wise, therefore approving, but not 100% sure about adding `invalidating` to it yet.

**Comment by madeleineostoja** [comment]
I can't think of a good API for this so take with a grain of salt, but IMO the only acceptable way for this to become part of navigation is if it was opt-in somehow. Requiring users to guard against `type !== 'invalidating'` in all their routing calls is asking for a footgun and even if you knew about it would get tiring real fast.

Rather than making it a `type` could it be an extra property in `Navigation`? Not as clean but much safer, ie:

```ts
interface Navigation {
  from: NavigationTarget;
  to: NavigationTarget;
  type: NavigationType;
  invalidation: boolean
  delta?: number;
}
```

It's weird that `type` would be null when invalidating then, but it's also weird that (presumably) `from` and `to` would also be null so 🤷🏻‍♀️

File: packages/kit/src/runtime/client/client.js [hunk 1]
```diff
@@ -144,11 +144,24 @@ export function create_client({ target, base, trailing_slash }) {
...
-				await update(url, []);
+				if (!cancelled) {
+					stores.navigating.set(navigation);
+					await update(url, []);
+					callbacks.after_navigate.forEach((cb) => cb(navigation));
+				}
 
 				invalidating = null;
 				force_invalidation = false;
+				stores.navigating.set(null);
```
Comment (PatrickG) [inline packages/kit/src/runtime/client/client.js:164]: Can this result in a wrong `navigation` state when the user clicks on a link while the invalidating is in progress?
Would probably only be noticeable on a slow connection.

**Comment by aradalvand** [comment]
I agree with @madeleineostoja, so what about something similar to [@lukaszpolowczyk's suggestion](https://github.com/sveltejs/kit/issues/6517#issuecomment-1235358605) where `beforeNavigate` and `afterNavigate` receive an `options` argument that lets the developer decide whether they want invalidation to be treated as navigation or not?
```js
beforeNavigate(({ type }) => {
    // show loader
}, { includeInvalidation: true });
```
@Rich-Harris Is this not feasible?

**Comment by madeleineostoja** [comment]
I think `$navigating` is less of an issue, since you're less likely to be using it to perform an action that you don't want an invalidation trigger.

I like that proposal. It's a little clunky, but much safer at the cost of a little verbosity. I'd name the option something shorter but that's neither here nor there.

**Comment by Rich-Harris** [comment]
I'm going to close this since the consensus (and I think I agree) seems to be against this solution. `$invalidating`, `beforeInvalidate`, `includeInvalidation` or any other API approach we settle on can be added in a non-breaking way later

**Pull request closed without merging**